src/_Private/ConditionState.php

```diff
@@ -0,0 +1,139 @@
+<?hh
+/*
+ *  Copyright (c) 2004-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the hphp/hsl/ subdirectory of this source tree.
+ *
+ */
+
+namespace HH\Lib\_Private;
+
+use type HH\Lib\Async\Condition;
+
+<<__Sealed(
+  NotStarted::class,
+  SyncResult::class,
+  AsyncResult::class,
+  Finished::class,
+)>>
+interface ConditionState<T> {
```

Comment:
- this API feels a lot like a `Promise`; would that be a better naming?
- this + the interfaces potentially feels like a good use case for `enum class` - would that make sensehere?